### PR TITLE
Shyrma helpers

### PR DESCRIPTION
**Pull request by shyrma**
profiling pool helpers and implementing of tf averaging in back prop for avgpool

### CODE REVIEW

File: libnd4j/include/ops/declarable/generic/convo/pooling3d_ops.cpp [hunk 1]
```diff
@@ -120,37 +120,7 @@ namespace nd4j {
 
             output->assign(0.0);
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/convo/pooling3d_ops.cpp]: Do we still need this assign?

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp [hunk 2]
```diff
@@ -55,20 +56,20 @@ namespace ops  {
...
-            int strideBS   = colStrides[0];
-            int strideColC = colStrides[1];
-            int strideKD   = colStrides[2];
-            int strideKH   = colStrides[3];
-            int strideKW   = colStrides[4];
-            int strideColD = colStrides[5];
-            int strideColH = colStrides[6];
-            int strideColW = colStrides[7];
+            Nd4jLong *colShapeOnly = shape::shapeOf(col.getShapeInfo());
+            Nd4jLong *colStrides   = shape::stride(col.getShapeInfo());
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: get back auto please

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp [hunk 3]
```diff
@@ -737,56 +768,62 @@ void ConvolutionUtils<T>::maxPool3dFrame(NDArray<T>& input, NDArray<T>& output,
...
-        /* loop over output */
-        int i, j, ti;
-        for (ti = 0; ti < oD; ti++) {
-            for (i = 0; i < oH; i++) {
-                for (j = 0; j < oW; j++){
+    const int iStride2 = iH * iW;
+    const int oStride2 = oH * oW;
+    const int iStride1 = iD * iStride2;
+    const int oStride1 = oD * oStride2;
+    const int iStride0 = iC * iStride1;
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: use auto or Nd4jLong here please. Convolution stuff was the reason behind int -> long migration, exactly due to possibility of getting strides > MAX_INT.

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp [hunk 4]
```diff
@@ -795,75 +832,78 @@ void ConvolutionUtils<T>::maxPool3dFrame(NDArray<T>& input, NDArray<T>& output,
...
-    int* indBuff = indices + indStride;
+void ConvolutionUtils<T>::maxPool3dIndices(NDArray<T>& input, int* indices, const int oD, const int oH, const int oW, const int kD, const int kH, const int kW, const int sD, const int sH, const int sW, const int pD, const int pH, const int pW, const int dD, const int dH, const int dW) {
+
+    T* inBuff  = input.getBuffer();  
 
+    const int bS = input.sizeAt(0);
     const int iC = input.sizeAt(1);
     const int iD = input.sizeAt(2);
     const int iH = input.sizeAt(3);
     const int iW = input.sizeAt(4);
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: same as above. Nd4jLong please.

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp [hunk 5]
```diff
@@ -1410,154 +1458,89 @@ void ConvolutionUtils<T>::vol2col(NDArray<T>& volume, NDArray<T>& columns, const
...
-    const int volStride3 = volume.stridesOf()[3];
-    const int volStride4 = volume.stridesOf()[4];    
+    const Nd4jLong colStride0 = columns.stridesOf()[0];
+    const Nd4jLong colStride1 = columns.stridesOf()[1];
+    const Nd4jLong colStride2 = columns.stridesOf()[2];
+    const Nd4jLong colStride3 = columns.stridesOf()[3];
+    const Nd4jLong colStride4 = columns.stridesOf()[4];
+    const Nd4jLong colStride5 = columns.stridesOf()[5];
+    const Nd4jLong colStride6 = columns.stridesOf()[6];
+    const Nd4jLong colStride7 = columns.stridesOf()[7];
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: 👍

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp [hunk 6]
```diff
@@ -1410,154 +1458,89 @@ void ConvolutionUtils<T>::vol2col(NDArray<T>& volume, NDArray<T>& columns, const
...
+                        
+                                    volDep = (-pD + kDep * dD) + colD*sD;
+                                    volRow = (-pH + kRow * dH) + colH*sH;
+                                    volCol = (-pW + kCol * dW) + colW*sW;
+                                        
+                                    col = colBuff + b*colStride0 + c*colStride1 + kDep*colStride2 + kRow*colStride3 + kCol*colStride4 + colD*colStride5 + colH*colStride6 + colW*colStride7;;
+                                    vol = volBuff + b*volStride0 + c*volStride1 + volDep*volStride2 + volRow*volStride3 + volCol*volStride4;
+                                                    
+                                    if (static_cast<unsigned>(volDep) >= static_cast<unsigned>(iD) || static_cast<unsigned>(volRow) >= static_cast<unsigned>(iH) || static_cast<unsigned>(volCol) >= static_cast<unsigned>(iW))
+                                        *col = 0.;
```
Comment (raver119) [inline libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: start with float in this case:

```
*col = static_cast<T>(0.f);
```

File: libnd4j/include/ops/declarable/generic/convo/pooling3d_ops.cpp (hunk 1 above)
Comment (shyrma) [reply libnd4j/include/ops/declarable/generic/convo/pooling3d_ops.cpp]: indeed, it is unnecessary there, helper does this job by itself

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp (hunk 2 above)
Comment (shyrma) [reply libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: done

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp (hunk 3 above)
Comment (shyrma) [reply libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: done

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp (hunk 4 above)
Comment (shyrma) [reply libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: done

File: libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp (hunk 6 above)
Comment (shyrma) [reply libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp]: done, (T)0.

**Pull request merged**